### ISSUE TITLE
Wait time for SQLite with NCBI taxonomy.

### ISSUE THREAD

**Issue by grabear** [biosql/biosql#24]
Does anyone have any experience using sqlite as a driver for loading NCBI taxonomy data into a BioSQL database?

I'm working on running the perl script for this and it's taking forever.  I've been running it for more than 96 hours and it's only at 169MB.  It's definitely still uploading, it's just incredibly slow.

In the past I thought I did this much faster, but maybe I'm wrong...

Here are the commands that I used:

```shell
$ sqlite3 NCBI_BIOSQLITE.db -echo < biosqldb-sqlite.sql
$ ./load_ncbi_taxonomy.pl --db_name NCBI_BioSQLite.db --driver SQLite --download true
```

**Comment by Asutu** [comment]
I'm also experiencing the same issue with SQLite, with more than 24h as for now...

My commands were:

```
sqlite3 nt.db < biosqldb-sqlite.sql
perl load_ncbi_taxonomy.pl --dbname nt.db --driver SQLite --download --directory ./taxdata
```

**Comment by hlapp** [comment]
Could this be an issue with the query planner? E.g., in older versions of PostgreSQL, one had to update the index statistics so that the query planner wouldn't be completely misled by the statistics from when tables were nearly empty.

**Comment by Asutu** [comment]
Hi hlapp,

I'm very new to BioSQL and SQL in general so not sure how to proceed with those steps. Looking at `load_ncbi_taxonomy.pl`, there's already the following lines ([L505](https://github.com/biosql/biosql/blob/6b3baed0fae48bbeb66ee1b663f56bc8d1194be9/scripts/load_ncbi_taxonomy.pl#L505)), which I think are already updating the index statistics?

```
} elsif($driver eq "SQLite") {
	print STDERR "\t... (vacuuming)\n" if $verbose;
	$dbh->do("VACUUM");
	$dbh->do("ANALYZE taxon");
}
```

Edit:

Actually, after commenting out the above lines, the whole process took <10 min to finish!
However, I'm not sure if one should "VACUUM" and "ANALYZE" after the script finishes or not. I guess this is more SQLite-related than BioSQL, but what could be drawbacks of not vacumm analyze the database?

**Comment by hlapp** [comment]
> Actually, after commenting out the above lines, the whole process took <10 min to finish!

Glad you tried this, that's exactly what I was going to suggest.

I'm not familiar with SQLite and their recent development roadmap. It sounds like this used to be necessary but in the latest release(s) is done concurrently (and well) by the database engine? I.e., one question is should this bypassed for versions of SQLite ≥ X but kept for earlier ones?

If the script runs faster with bypassing the `VACUUM ANALYZE` then it's probably not necessary to run it at the end either. In other words, it is better left to the user (or their DBA) whether this should be run at the end or not.

**Comment by grabear** [comment]
Nice guys!  Thanks for looking into and responding here.  I will try this as well.

@hlapp What does `VACUUM ANALYZE` mean?

**Comment by grabear** [comment]
Can confirm that my SQLite database is loaded and the perl script has finished running.  It took ~30 minutes total to complete.

Any insight into this fix would be great.  Does a PR need to be made?

**Comment by hlapp** [comment]
> Any insight into this fix would be great. Does a PR need to be made?

Yes and yes. From my perspective, to include this in a release we'd have to clarify the question whether removing those commands will be disastrous for performance for earlier versions of SQLite. If yes, then it'd be great if can keep the commands but execute only conditionally on version being affected.

**Comment by taltman** [comment]
Former BioWarehouse developer here (http://bioinformatics.ai.sri.com/biowarehouse/), jumping into the fray...

@hlapp: Where did you read about the SQLite VACUUM command not being needed in recent versions?

@grabear: The VACUUM command (in both PostgreSQL and SQLite) deals with database table & index fragmentation under transactional use-cases where there are lots of row deletions and insertions. 

This would come up with the taxon table in BioSQL when using the load_ncbi_taxonomy.pl script is used to update an existing table. It should not be necessary to use it *at all* if you are loading the NCBI Taxonomy DB from scratch. Using the command to keep obsoleted taxa should reduce fragmentation, but obsolete taxon names are still deleted, AFAICT.

The ANALYZE command in SQLite updates statistics that the query planner / DB engine uses. It is recommended when the database has changed significantly. Again, this should only be needed when doing a major update to the taxon table. 

In general, I think that this script could be made much more streamlined if one just takes a data warehousing approach. By that, I mean where you upload the NCBI Taxonomy DB from scratch each time, and different versions of the same database can coexist. I think the update use-case greatly complicates things (my $0.02).

I would advocate for a quick fix by making the execution of the VACUUM and ANALYZE commands a command-line option, so that people who need it can have it, but it isn't executed by default (especially in the build-from-scratch use-case). If I have time, I might issue a PR for that. Thanks!

**Comment by sachalau** [comment]
Hello everyone,

I'm using the load_ncbi_taxonomy.pl script to... load the NCBI taxonomy on an empy PostgreSQL BioSQL instance.

I've noticed the very long run time (3h30 approximately). 
I dont remember the breaking up of the time for each step, but currently "... insert / update / delete taxon nodes" has been running for one hour. 
The other step that takes quite a large amount of time is the building of left/right values.

As this is use case is for charging an empty database, do you think I could tweak the input parameters to speed up things initially ?

I would definitely submit a PR if that were the case... and knew enough Perl.

Thanks for your help.
